File: apps/mobile/modules/db-native/ios/DbNativeModule.swift
import ExpoModulesCore

// Thin bridge between the JS db-native module and the shared-rust crate, generated as a
// Swift binding by `uniffi-bindgen` (see ios/generated/shared_rust.swift).
//
// NOT built/verified on this (Windows) machine — there is no Xcode/macOS toolchain available.
// To finish wiring this up on macOS:
//   1. Cross-compile shared-rust for iOS targets, e.g. via `cargo build --release --target aarch64-apple-ios`
//      (and `aarch64-apple-ios-sim` / `x86_64-apple-ios` for simulators), then package the
//      resulting libraries into an XCFramework.
//   2. Add the XCFramework to ios/ and reference it from db-native.podspec.
//   3. `cd apps/mobile/ios && pod install`, then build via Xcode / `expo run:ios`.
public class DbNativeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DbNative")

    AsyncFunction("connect") { (dbType: String, connectionString: String) -> String in
      try connect(dbType: dbType, connectionString: connectionString)
    }

    AsyncFunction("disconnect") { (connectionId: String) -> Void in
      try disconnect(connectionId: connectionId)
    }

    AsyncFunction("executeQuery") { (connectionId: String, query: String) -> [String: Any] in
      let result = try executeQuery(connectionId: connectionId, query: query)
      return [
        "rows": result.rows,
        "affectedRows": result.affectedRows
      ]
    }

    AsyncFunction("getSchema") { (connectionId: String) -> [String: Any] in
      let schema = try getSchema(connectionId: connectionId)
      return [
        "tables": schema.tables.map { table in
          [
            "name": table.name,
            "columns": table.columns.map { col in
              [
                "name": col.name,
                "dataType": col.dataType,
                "isPrimary": col.isPrimary,
                "isNullable": col.isNullable
              ]
            }
          ]
        }
      ]
    }
  }
}

File: apps/mobile/modules/db-native/android/src/main/java/expo/modules/dbnative/DbNativeModule.kt
package expo.modules.dbnative

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import uniffi.shared_rust.DbException
import uniffi.shared_rust.connect as rustConnect
import uniffi.shared_rust.disconnect as rustDisconnect
import uniffi.shared_rust.executeQuery as rustExecuteQuery
import uniffi.shared_rust.getSchema as rustGetSchema

// Thin bridge between the JS db-native module and the shared-rust crate, generated as a
// JNA-backed Kotlin binding by `uniffi-bindgen` (see android/src/main/java/uniffi/shared_rust).
class DbNativeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DbNative")

    AsyncFunction("connect") { dbType: String, connectionString: String ->
      try {
        rustConnect(dbType, connectionString)
      } catch (e: DbException) {
        throw DbNativeException(e)
      }
    }

    AsyncFunction("disconnect") { connectionId: String ->
      try {
        rustDisconnect(connectionId)
      } catch (e: DbException) {
        throw DbNativeException(e)
      }
    }

    AsyncFunction("executeQuery") { connectionId: String, query: String ->
      try {
        val result = rustExecuteQuery(connectionId, query)
        mapOf(
          "rows" to result.rows,
          "affectedRows" to result.affectedRows.toLong()
        )
      } catch (e: DbException) {
        throw DbNativeException(e)
      }
    }

    AsyncFunction("getSchema") { connectionId: String ->
      try {
        val schema = rustGetSchema(connectionId)
        mapOf(
          "tables" to schema.tables.map { table ->
            mapOf(
              "name" to table.name,
              "columns" to table.columns.map { col ->
                mapOf(
                  "name" to col.name,
                  "dataType" to col.dataType,
                  "isPrimary" to col.isPrimary,
                  "isNullable" to col.isNullable
                )
              }
            )
          }
        )
      } catch (e: DbException) {
        throw DbNativeException(e)
      }
    }
  }
}

private fun DbNativeException(e: DbException): expo.modules.kotlin.exception.CodedException {
  val message = when (e) {
    is DbException.Generic -> e.details
  }
  return expo.modules.kotlin.exception.CodedException(message, e)
}
